wasm-text.md

```diff
@@ -360,6 +360,229 @@ Imports can be declared like regular functions, memories, etc., by giving an inl
+    rule #t2aInstr<C>( if  ID:Identifier TDS IS else OID':OptionalId IS' end OID'') =>  if  ID TDS #t2aInstrs<C>(IS) else OID' #t2aInstrs<C>(IS') end OID''
+    rule #t2aInstr<C>( if OID:OptionalId TDS IS                          end OID'') =>  if OID TDS #t2aInstrs<C>(IS)                              end OID''
+    rule #t2aInstr<C>((if OID:OptionalId TDS COND (then IS)))                       => (if OID TDS #t2aInstrs<C>(COND) (then #t2aInstrs<C>(IS)))
+    rule #t2aInstr<C>((if OID:OptionalId TDS COND (then IS) (else IS')))            => (if OID TDS #t2aInstrs<C>(COND) (then #t2aInstrs<C>(IS)) (else #t2aInstrs<C>(IS')))
+```
+
+#### KWasm Administrative Instructions
+
+The following instructions are not part of the official Wasm text format.
+They are currently supported in KWasm text files, but may be deprecated.
+
+```k
+    rule #t2aInstr<C>(trap) => trap
+
+    rule #t2aInstr<C>(block VT:VecType IS:Instrs end) => block VT #t2aInstrs<C>(IS) end
+
+    rule #t2aInstr<_>(init_local I V) => init_local I V
+    rule #t2aInstr<_>(init_locals VS) => init_locals VS
+```
+
+### List Functions
+
+The following are helper functions.
+They distribute the text-to-abstract functions above over lists.
+
+```k
+    syntax Stmts      ::= "#t2aStmts"      "<" Context ">" "(" Stmts      ")" [function]
+    syntax Defns      ::= "#t2aDefns"      "<" Context ">" "(" Defns      ")" [function]
+    syntax Instrs     ::= "#t2aInstrs"     "<" Context ">" "(" Instrs     ")" [function]
+    syntax LocalDecls ::= "#t2aLocalDecls" "<" Context ">" "(" LocalDecls ")" [function]
+ // ------------------------------------------------------------------------------------
+    rule #t2aStmts<C>(S:Stmt SS:Stmts) => #t2aStmt<C>(S) #t2aStmts<C>(SS)
+    rule #t2aStmts<_>(.Stmts) => .Stmts
+
+    rule #t2aDefns<C>(D:Defn DS:Defns) => #t2aDefn<C>(D) #t2aDefns<C>(DS)
+    rule #t2aDefns<_>(.Defns) => .Defns
+
+    rule #t2aInstrs<C>(I:Instr IS:Instrs) => #t2aInstr<C>(I) #t2aInstrs<C>(IS)
+    rule #t2aInstrs<_>(.Instrs) => .Instrs
+
+    rule #t2aLocalDecls<C>(LD:LocalDecl LDS:LocalDecls) => #t2aLocalDecl<C>(LD) #t2aLocalDecls<C>(LDS)
+    rule #t2aLocalDecls<_>(.LocalDecls) => .LocalDecls
+```
+
+### Functions for Gathering Context
+
+The following are helper functions for gathering and updating context.
+
+```k
+    syntax Map        ::= #ids2Idxs(TypeUse, LocalDecls)      [function, functional]
+                        | #ids2Idxs(Int, TypeUse, LocalDecls) [function, functional]
+ // --------------------------------------------------------------------------------
+    rule #ids2Idxs(TU, LDS) => #ids2Idxs(0, TU, LDS)
+
+    rule #ids2Idxs(_, .TypeDecls, .LocalDecls) => .Map
+    rule #ids2Idxs(N, (type _)    , LDS) => #ids2Idxs(N, .TypeDecls, LDS)
+    rule #ids2Idxs(N, (type _) TDS, LDS) => #ids2Idxs(N, TDS       , LDS)
+
+    rule #ids2Idxs(N, (param ID:Identifier _) TDS, LDS)
+      => (ID |-> N) #ids2Idxs(N +Int 1, TDS, LDS)
```

Comment:
You might prefer the syntax `#ids2Idxs(N +Int 1, TDS, LDS) [ ID <- N ]`, which means something different than what you have written.

What you have written will halt with a K runtime error if the same `ID` is entered into the map twice (and so we're attempting a map union over two maps with the same key). The example I wrote will continue executing, but will prefer later definitions for the same identifier.

Not sure if it's a valid thing to have the same identifier show up multiple times, but something for you to consider.

Reply:
The same identifier can't appear twice.

> An identifier context is well-formed if no index space contains duplicate identifiers.
https://webassembly.github.io/spec/core/text/conventions.html#contexts

So runtime error on transforming to abstract syntax seems more appropriate than overwriting. Again, we could exit abnormally with some message.